### DISCUSSION TITLE
Add cherryusb as a module

### DISCUSSION THREAD

**Discussion by sakumisu** [zephyrproject-rtos/zephyr#49108] [General]
Hi,are you interested in cherryusb? I wanted to contribute it to zephyr for using usb device and host if it is possible. The link is here: [CherryUSB](https://github.com/sakumisu/CherryUSB). If you are interested, please inform me, thank you!

**Comment by stephanosio** [comment]
FYI @jfischer-no

**Reply by jfischer-no** [reply]
> Hi,are you interested in cherryusb? I wanted to contribute it to zephyr for using usb device and host if it is possible. The link is here: [CherryUSB](https://github.com/sakumisu/CherryUSB). If you are interested, please inform me, thank you!

No, I do not think so. We already have USB device support in Zephyr and also working on the host support, see https://github.com/zephyrproject-rtos/zephyr/issues/42066, you are welcome to contribute to this work.

**Reply by sakumisu** [reply]
Well, my stack has many advantages in both speed and memory, and with many ports support. I am very familiar with many usb ips and my job is to be responsible for usb ip design. Finally, many companies are already using it. If it is not possible, i will checkout zephyr as cherryusb demo because i have some partners using zephyr with cherryusb. Thank you for your reply.

**Reply by stephanosio** [reply]
> my stack has many advantages in both speed and memory

without some numbers and tables comparing the "CherryUSB" and the Zephyr's own USB stack, this won't get far.

**Reply by sakumisu** [reply]
> 

I can't do the comparison with others on my side, just for mine, you can see code size in my stack. And all my ports are based on registers, speed is the limit of hardware usb ip. And very friendly for newbies, you can get started in three minutes. It's okay, it doesn't matter, just your choice.

**Reply by jfischer-no** [reply]
> Well, my stack has many advantages in both speed and memory, and with many ports support.

I doubt that very much, especially device support part. I have been working with Zephyr's USB code for a few years now, and when I look at the device part of CherryUSB, there are parts that look very familiar, which is very strange. I am speculating now, and I may be wrong, but I have the impression that it is a kind of copy of the current Zephyr device support (for which there is already successor), which is based on (LPCUSB, Copyright (C) 2006 Bertrik Sikken (...)).  Alone when one looks at https://github.com/sakumisu/CherryUSB/blob/master/core/usbd_core.c, it is very similar to ours in many places, probably on tree state a year ago, we had security fixes that I cannot find in CherryUSB.

>I am very familiar with many usb ips and my job is to be responsible for usb ip design. Finally, many companies are already using it.

Well, you know "finally, many companies are already using" Zephyr's USB support.

>If it is not possible, i will checkout zephyr as cherryusb demo because i have some partners using zephyr with cherryusb. Thank you for your reply.

In general, there will be no approval from my side for a USB support as module, this must be a native implementation, and also not what looks like a copy of what we already have.

**Reply by jfischer-no** [reply]
Very interesting https://github.com/sakumisu/CherryUSB/commit/8a143df5094b0b1db72c8b9c3120cfc390b54c50, use of Z_ prefix must be catching ("Use Z_MAX"), also USB device controller API, with all the issues and weaknesses we fix in overhauled USB device support. It can all be a coincidence of course, I apologize then.

**Reply by jfischer-no** [reply]
> Yes, some codes about enumration refers to zephyr.Ok, i will continuously be attention to zephyr later plan.Later, i wish my ports can give some references to you when you implement new udc or hcd.

No sir, it is not just that, and anyway what happened to Copyright (C) 2006 Bertrik Sikken and others? I do not think I will welcome your contribution to Zephyr code.

**Reply by carlescufi** [reply]
USB support has to be core to Zephyr, it cannot live in a separate repo, it's too important/core functionality. So the options you have are:

- Fork Zephyr, include cherryusb in your fork
- Advertise cherryusb as a Zephyr module that users can pull into their manifests
- Contribute cherryusb to Zephyr

**Reply by sakumisu** [reply]
> USB support has to be core to Zephyr, it cannot live in a separate repo, it's too important/core functionality. So the options you have are:
> 
> * Fork Zephyr, include cherryusb in your fork
> * Advertise cherryusb as a Zephyr module that users can pull into their manifests
> * Contribute cherryusb to Zephyr

Thank you, the first two I am working on.

**Comment by jfischer-no** [comment]
Besides a lot of unique parts, one simple copy&paste error was forgotten right at the beginning :-), 
https://github.com/sakumisu/CherryUSB/commit/8a143df5094b0b1db72c8b9c3120cfc390b54c50#diff-2bfcf3121e84c376e4962b654d18f42fe08f86303624da5e067bfba63870731aR302

![49108_comment](https://user-images.githubusercontent.com/9864546/184920672-8a62ca4d-e652-4c35-865f-309b7eb0de07.png)

`git log -S zephyr` is also interesting, I try to find more time to investigate git tree.
